File: 2078. two-furthest-houses-with-different-colors.java
class Solution {
    static public int maxDistance(int[] colors) {
        final int n=colors.length;
        final int c0=colors[0], cN=colors[n-1];
        int lMax=0, rMax=0;
        for(int i=0; i<n; i++){
            final int c=colors[i];
            if (c0!=c && i>lMax) lMax=i;
            if (cN!=c && n-1-i>rMax) rMax=n-1-i;
        }
        return Math.max(lMax, rMax);
    }
    
    public static void main(String[] args) {
        // Example 1
        int[] colors1 = {1,1,1,6,1,1,1};
        int output1 = maxDistance(colors1);
        System.out.println("Example 1: colors=[1,1,1,6,1,1,1]");
        System.out.println("Output: " + output1);
        System.out.println("Expected: 3");
        System.out.println();
        
        // Example 2
        int[] colors2 = {1,8,3,8,3};
        int output2 = maxDistance(colors2);
        System.out.println("Example 2: colors=[1,8,3,8,3]");
        System.out.println("Output: " + output2);
        System.out.println("Expected: 4");
        System.out.println();
        
        // Example 3
        int[] colors3 = {0,1};
        int output3 = maxDistance(colors3);
        System.out.println("Example 3: colors=[0,1]");
        System.out.println("Output: " + output3);
        System.out.println("Expected: 1");
    }
}

File: 2078. two-furthest-houses-with-different-colors.py
from typing import List

class Solution:
    def maxDistance(self, colors: List[int]) -> int:
        n=len(colors)
        c0, cN=colors[0], colors[-1]
        lMax, rMax=0, 0
        for i, c in enumerate(colors):
            if c0!=c: lMax=max(lMax, i)
            if cN!=c: rMax=max(rMax, n-1-i)
        return max(lMax, rMax)

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1
    colors1 = [1,1,1,6,1,1,1]
    output1 = sol.maxDistance(colors1)
    print(f"Example 1: colors={colors1}")
    print(f"Output: {output1}")
    print(f"Expected: 3")
    print()
    
    # Example 2
    colors2 = [1,8,3,8,3]
    output2 = sol.maxDistance(colors2)
    print(f"Example 2: colors={colors2}")
    print(f"Output: {output2}")
    print(f"Expected: 4")
    print()
    
    # Example 3
    colors3 = [0,1]
    output3 = sol.maxDistance(colors3)
    print(f"Example 3: colors={colors3}")
    print(f"Output: {output3}")
    print(f"Expected: 1")